### PR TITLE
Typed Constraints

### PR DESCRIPTION
**Pull request by vvtran**
- https://github.com/Plutonomicon/cardano-browser-tx/issues/166
- Added typed constraints to script lookups using functional deps instead of associated types which aren't available in Purescript.
- Added a few easy `ToData` `FromData` instances.
- Updated script lookups to include these type constraints.

### CODE REVIEW

File: src/Types/Any.purs [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+  ) where
+
+import Prelude
+import Data.Generic.Rep (class Generic)
+import Data.Show.Generic (genericShow)
+
+-- Not sure if this is necessary but keep for now.
+-- | For converting a subset of polymorphic types to that which will work with
+-- | `PlutusData`.
+data Any
```
Comment (ngua) [inline src/Types/Any.purs:12]: Missing a kind signature

File: src/Types/Transaction.purs [hunk 2]
```diff
@@ -70,6 +73,14 @@ instance monoidTransaction :: Monoid Transaction where
     , auxiliary_data: Nothing
     }
 
+-- Do we actually need this?
```
Comment (ngua) [inline src/Types/Transaction.purs:76]: I really don't understand this instance or the one below. I can't imagine how it could be used sensibly

File: src/Types/TypedTxOut.purs [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+  let address = typedValidatorAddress networkId typedVal
+  pure $ mkTypedTxOut' (wrap { address, amount, data_hash: pure dHash }) dt
+  where
+  mkTypedTxOut'
+    :: TransactionOutput
+    -> b -- Data
+    -> TypedTxOut a b
+  mkTypedTxOut' txOut dat = TypedTxOut { txOut, data: dat }
+
+data WrongOutTypeError = ExpectedScriptGotPubkey -- | ExpectedPubkeyGotScript
```
Comment (ngua) [inline src/Types/TypedTxOut.purs:179]: What's going on with the commented-out constructor at the end?

File: src/Types/TypedTxOut.purs [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+
+derive instance Generic WrongOutTypeError _
+derive instance Eq WrongOutTypeError
+derive instance Ord WrongOutTypeError
+
+instance Show WrongOutTypeError where
+  show = genericShow
+
+-- | An error we can get while trying to type an existing transaction part.
+data ConnectionError
```
Comment (ngua) [inline src/Types/TypedTxOut.purs:189]: Is this name taken from Plutus? If so, I don't think we should follow it. Especially in the context of CTL, I think the name would be take to mean a failure in some networked code

File: src/Types/TypedTxOut.purs [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+  -> Address
+  -> m (Either ConnectionError Unit)
+checkValidatorAddress networkId typedVal actualAddr = runExceptT do
+  let expectedAddr = typedValidatorAddress networkId typedVal
+  unless (expectedAddr == actualAddr)
+    $ throwError
+    $ WrongValidatorAddress expectedAddr actualAddr
+
+-- -- | Checks that the given redeemer script has the right type.
+-- checkRedeemer
```
Comment (ngua) [inline src/Types/TypedTxOut.purs:222]: Do we need this? Is this related to the note on `addOwnInput`?

File: src/Types/TypedTxOut.purs [hunk 6]
```diff
@@ -0,0 +1,289 @@
+module Types.TypedTxOut
```
Comment (ngua) [inline src/Types/TypedTxOut.purs:1]: This whole module is to make the typed validators possible and there is no analog in Plutus since it's not necessary, right? If so, can we add a doc comment to the module explaining this?

File: src/Types/TypedValidator.purs [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+typedValidatorScript = _.validator <<< unwrap
+
+-- | The minting policy that forwards all checks to the instance's validator.
+forwardingMintingPolicy
+  :: forall (a :: Type). TypedValidator a -> MintingPolicy
+forwardingMintingPolicy = _.forwardingMPS <<< unwrap
+
+-- We have a few functions, I'm not sure if we even need these for off chain
+-- code:
+-- -- Broken, see below (we need some notion of `applyCode`)
```
Comment (ngua) [inline src/Types/TypedValidator.purs:130]: We will have a server call that applies `PlutusData` arguments to a `Script` soon, but that is the only type of `applyCode` that we will be able to support (and it will be in `QueryM`)

File: src/Types/Transaction.purs (hunk 2 above)
Comment (vvtran) [reply src/Types/Transaction.purs:76]: It's used in `wrapValidator` constraints, although I'm not convinced we need it ATM. I could comment this function out for now?

File: src/Types/TypedTxOut.purs (hunk 3 above)
Comment (vvtran) [reply src/Types/TypedTxOut.purs:179]: There seems to be a separation of checking whether something is a PubKey vs Script which is used in the `TxIn` version of our `typeTxOutRef`. https://playground.plutus.iohkdev.io/doc/haddock/plutus-ledger/html/src/Ledger.Typed.Tx.html#PubKeyTxOut

I'm tempted to juse remove this datatype and add `ExpectedScriptGotPubkey` as an option to `ConnectionError`

File: src/Types/TypedTxOut.purs (hunk 4 above)
Comment (vvtran) [reply src/Types/TypedTxOut.purs:189]: It is, I will rename.

File: src/Types/TypedTxOut.purs (hunk 5 above)
Comment (vvtran) [reply src/Types/TypedTxOut.purs:222]: Yes, it's related to the fact they use `TxIn` in `addOwnInput` which adds a redeemer which we don't. I can remove if you want.

File: src/Types/TypedTxOut.purs (hunk 6 above)
Comment (vvtran) [reply src/Types/TypedTxOut.purs:1]: https://playground.plutus.iohkdev.io/doc/haddock/plutus-ledger/html/src/Ledger.Typed.Tx.html This is the equivalent module. I can add more comments.

File: src/Types/Any.purs (hunk 1 above)
Comment (vvtran) [reply src/Types/Any.purs:12]: There's no polymorphic variable here and I don't think we have standalone kinds, do we, if that's what you are referring to?

Comment (vvtran) [reply src/Types/Any.purs]: Just implemented some changes if you could check.

**Review by ngua — APPROVED** [review]
